File: Data Structures & Algorithms/string-encode-and-decode/submission-1.py
class Solution:

    def encode(self, strs: List[str]) -> str:
        res = [] 
        for s in strs: 
            for char in s:
                if char == '1' or char == '0': 
                    res.append('0') 
                res.append(char) 
            res.append('1')
        return ''.join(res) 

    def decode(self, s: str) -> List[str]:
        res = [] 
        temp = [] 

        i = 0 
        while i < len(s): 
            char = s[i] 

            if char == '0': 
                temp.append(s[i+1])
                i += 1 
            elif char == '1': 
                res.append(''.join(temp)) 
                temp = [] 
            else:
                temp.append(char)
            i += 1  
        return res 

File: Data Structures & Algorithms/anagram-groups/submission-1.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        map<array<int, 26>, vector<string>> mp; 

        for(const string& s: strs){
            array<int, 26> freq = getFreq(s); 
            mp[freq].push_back(s); 
        }

        vector<vector<string>> res; 
        for(auto const& [key, val]: mp){
            res.push_back(val);
        }

        return res;
    }

    array<int, 26> getFreq(const string& s){
        array<int, 26> freq = {};
        
        for(auto ch: s){
            freq[ch - 'a']++;
            // int ord = (int)ch - 97;
            // freq[ord]++;
        }
        return freq; 
    }
};
